contracts/pcv/EthPCVSwapperUniswap.sol

```diff
@@ -0,0 +1,145 @@
+    uint256 public maxEthSpentPerSwap;
+    /// @notice the maximum amount of slippage vs oracle price
+    uint256 public maximumSlippage = 300; // default 3%
+    uint256 public constant SLIPPAGE_GRANULARITY = 10000;
+
+    constructor(
+        address _core,
+        address _pair,
+        address _router,
+        address _oracle,
+        uint256 _swapFrequency,
+        address _tokenReceived,
+        uint256 _tokenBuyLimit,
+        uint256 _maxEthSpentPerSwap
+    ) public UniRef(_core, _pair, _router, _oracle) Timed(_swapFrequency) {
+        tokenReceived = _tokenReceived;
+        tokenBuyLimit = _tokenBuyLimit;
+        maxEthSpentPerSwap = _maxEthSpentPerSwap;
+
+        emit UpdateTokenSpent(router.WETH());
+        emit UpdateTokenReceived(_tokenReceived);
+
+        // start timer
+        _initTimed();
+    }
+
+    receive() external payable {}
+
+    /// @notice withdraw ETH from the reserves
+    /// @param to address to send ETH
+    /// @param amountOut amount of ETH to send
+    function withdrawETH(address payable to, uint256 amountOut) external override onlyPCVController {
+        Address.sendValue(to, amountOut);
+        emit WithdrawETH(msg.sender, to, amountOut);
+    }
+
+    /// @notice withdraw ERC20 from the reserves
+    /// @param to address destination of the ERC2°
+    /// @param token address of the ERC20 to send
+    /// @param amount amount of ERC20 to send
+    function withdrawERC20(address to, address token, uint256 amount) external override onlyPCVController {
+        require(IERC20(token).transfer(to, amount), "EthPCVSwapperUniswap: transferERC20 failed");
+        emit WithdrawERC20(msg.sender, to, token, amount);
+    }
+
+    /// @notice Sets the token to spend
+    /// @param _tokenSpent the address of the token to spend
+    function setTokenSpent(address _tokenSpent) external override onlyGovernor {
+        require(_tokenSpent == router.WETH(), "EthPCVSwapperUniswap: can only spend ETH");
+    }
+
+    /// @notice Sets the token to receive
+    /// @param _tokenReceived the address of the token to receive
+    function setTokenReceived(address _tokenReceived) external override onlyGovernor {
+      tokenReceived = _tokenReceived;
+      emit UpdateTokenReceived(_tokenReceived);
+    }
+
+    /// @notice Get the token to spend
+    /// @return The address of the token to spend
+    function getTokenSpent() external view override returns (address) {
+      return router.WETH();
+    }
+
+    /// @notice Get the token to receive
+    /// @return The address of the token to receive
+    function getTokenReceived() external view override returns (address) {
+      return tokenReceived;
+    }
+
+    function swap() external override afterTime whenNotPaused {
+      require(address(this).balance > 0, "EthPCVSwapperUniswap: no ETH left");
+      require(IERC20(tokenReceived).balanceOf(address(this)) < tokenBuyLimit, "EthPCVSwapperUniswap: cannot buy more");
```

Comment:
We could end up going ~maxEthSpentPerSwap beyond tokenBuyLimit, we could consider requiring after the trade or even removing this condition and having the intended usage be a full conversion between assets

Reply:
I think that's acceptable? Just like the `EthPCVDripper` dripping chunks of 5000 ETH can put the target's balance in the [10000, 5000] range? The idea is just to have a backstop if desired, otherwise we can set it to `UINT256_MAX` for full conversion.

If each swap is ~0.1% of the target amount, worst case we'll end up with 100.1% of our target amount of the target ERC20, which is a smaller noise than the potential slippage.

Reply:
I've set a default value of 0 (meaning : unlimited) in the constructor, with a setter to define a limit if desired.